rasa/core/training/story_conflict.py

```diff
@@ -0,0 +1,221 @@
+        hashed_state = hash(str(list(element.sliced_states)))
+        if element.event.as_story_string() not in state_action_dict[hashed_state]:
+            state_action_dict[hashed_state] += [element.event.as_story_string()]
+
+    # Keep only conflicting `state_action_dict`s
+    return {
+        state_hash: actions
+        for (state_hash, actions) in state_action_dict.items()
+        if len(actions) > 1
+    }
+
+
+def _build_conflicts_from_states(
+    trackers: List["TrackerWithCachedStates"],
+    domain: Domain,
+    max_history: int,
+    rules: Dict[int, Optional[List[Text]]],
+) -> List["StoryConflict"]:
+    # Iterate once more over all states and note the (unhashed) state,
+    # for which a conflict occurs
+    conflicts = {}
+    for element in _sliced_states_iterator(trackers, domain, max_history):
+        hashed_state = hash(str(list(element.sliced_states)))
+
+        if hashed_state in rules and hashed_state not in conflicts:
+            conflicts[hashed_state] = StoryConflict(element.sliced_states)
+
+        if hashed_state in rules:
+            conflicts[hashed_state].add_conflicting_action(
+                action=element.event.as_story_string(),
+                story_name=element.tracker.sender_id,
+            )
+
+    # Remove conflicts that arise from unpredictable actions
+    return [
+        conflict
+        for (hashed_state, conflict) in conflicts.items()
+        if conflict.has_prior_events
+    ]
+
+
+def _sliced_states_iterator(
+    trackers: List[TrackerWithCachedStates], domain: Domain, max_history: int
+) -> TrackerEventStateTuple:
+    """Creates an iterator over sliced states.
+
+    Iterate over all given trackers and all sliced states within each tracker,
+    where the slicing is based on `max_history`.
+
+    Args:
+        trackers: List of trackers.
+        domain: Domain (used for tracker.past_states).
+        max_history: Assumed `max_history` value for slicing.
+
+    Yields:
+        A (tracker, event, sliced_states) triplet.
+    """
+    for tracker in trackers:
+        states = tracker.past_states(domain)
+        states = [dict(state) for state in states]
+
+        idx = 0
+        for event in tracker.events:
+            if isinstance(event, ActionExecuted):
+                sliced_states = MaxHistoryTrackerFeaturizer.slice_state_history(
+                    states[: idx + 1], max_history
+                )
+                yield TrackerEventStateTuple(tracker, event, sliced_states)
+                idx += 1
+
+
+def _get_previous_event(
+    state: Optional[Dict[Text, float]]
+) -> [Optional[Text], Optional[Text]]:
+    """Returns previous event type and name.
+
+    Returns the type and name of the event (action or intent) previous to the
+    given state.
+
+    Args:
+        state: Element of sliced states.
+
+    Returns:
+        Tuple of (type, name) strings of the prior event.
+    """
+    prev_event_type = None
```

Comment:
let's spell this out :-)
```suggestion
    previous_event_type = None
```